### PR TITLE
Node and Workload API support for federated bundles

### PR DESCRIPTION
**Pull request by azdagron**
This PR updates the Node API to return bundles with x509 and JWT SVID updates. The agent has been updated to store all bundles returned from x509 SVID updates in its cache. JWT support in the agent has not been implemented and handling the bundles will be done at that time.

The workload API has been extended to return a map of bundles with updates as well as a list of bundle ids that each workload is associated with.

### CODE REVIEW

File: pkg/agent/manager/manager.go [hunk 1]
```diff
@@ -39,7 +39,7 @@ type Manager interface {
 
 	// SubscribeToBundleChanges returns a new observer.Stream on which []*x509.Certificate instances are
```
Comment (MarcosDY) [inline pkg/agent/manager/manager.go]: it is no longer returning a observer.Stream

File: pkg/server/endpoints/node/handler_test.go [hunk 2]
```diff
@@ -650,8 +671,18 @@ func getExpectedFetchX509SVID(data *fetchSVIDData) *node.X509SVIDUpdate {
 	caCert, _, _ := util.LoadCAFixture()
 	svidUpdate := &node.X509SVIDUpdate{
 		Svids:               svids,
-		Bundle:              caCert.Raw,
+		DEPRECATEDBundle:    caCert.Raw,
```
Comment (MarcosDY) [inline pkg/server/endpoints/node/handler_test.go]: when we'll remove those deprecated fields? must we add some anotations/comments about a prossible version where we'll remove? or at leas when it was mark as deprecated? so we can track how long we have it?

File: proto/api/node/node.proto [hunk 3]
```diff
@@ -26,12 +35,18 @@ message X509SVIDUpdate {
     // keys. Map[SPIFFE_ID] => SVID.
     map<string, X509SVID> svids = 1;
 
-    // Latest SPIRE Server bundle
-    bytes bundle = 2;
+    // DEPRECATED. Latest SPIRE Server bundle.
```
Comment (MarcosDY) [inline proto/api/node/node.proto]: can you add when it was deprecated (version)? to make it easier to track

File: pkg/agent/client/client.go [hunk 4]
```diff
@@ -138,12 +143,14 @@ func (c *client) FetchUpdates(ctx context.Context, req *node.FetchX509SVIDReques
 		for spiffeid, svid := range resp.SvidUpdate.Svids {
 			svids[spiffeid] = svid
 		}
-		lastBundle = resp.SvidUpdate.Bundle
+		for spiffeid, bundle := range resp.SvidUpdate.Bundles {
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/client/client.go]: Yes, bug ranging over a nil map is not an error. It acts as if the map is empty.

File: pkg/agent/client/client.go [hunk 5]
```diff
@@ -129,7 +129,12 @@ func (c *client) FetchUpdates(ctx context.Context, req *node.FetchX509SVIDReques
 		}
 		if err != nil {
 			// There was an error receiving a response, exit loop to return what we have.
-			return &Update{regEntries, svids, lastBundle}, err
+			logrus.Errorf("failed to consume entire SVID update stream: %v", err)
+			return &Update{
+				Entries: regEntries,
+				SVIDs:   svids,
+				Bundles: bundles,
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/client/client.go]: Good catch. The old code returned aggregated update results and the error. However, the only consumer of the function does not use the returned update if there is an error. I'm not sure why the old code returned it. I think I'd rather `return nil, err` in this case.

File: pkg/agent/manager/cache/cache.go [hunk 6]
```diff
@@ -35,17 +81,15 @@ type Cache interface {
 	DeleteEntry(regEntry *common.RegistrationEntry) bool
 	// Entries returns all the in force cached entries.
 	Entries() []*Entry
-	// IsEmpty returns true if this cache doesn't have any entry.
-	IsEmpty() bool
 	// Registers and returns a Subscriber, and then sends latest WorkloadUpdate on its channel
 	Subscribe(selectors Selectors) Subscriber
 	// Set the bundle
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: 👍

File: pkg/agent/manager/cache/cache.go [hunk 7]
```diff
@@ -35,17 +81,15 @@ type Cache interface {
...
-	SetBundle([]*x509.Certificate)
-	// Retrieve the bundle
+	SetBundles(map[string][]*x509.Certificate)
+	// Retrieve the bundle for the trust domain
 	Bundle() []*x509.Certificate
 	// SubscribeToBundleChanges returns a new observer.Stream of []*x509.Certificate instances. Each
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: 👍

File: pkg/agent/manager/manager.go (hunk 1 above)
Comment (azdagron) [reply pkg/agent/manager/manager.go]: 👍

File: pkg/server/endpoints/node/handler_test.go (hunk 2 above)
Comment (azdagron) [reply pkg/server/endpoints/node/handler_test.go]: Hmm, i think that is a larger discussion to have about API deprecation paths. For now, is git history sufficient to know when it was deprecated?

**Pull request merged**